server/src/main/java/org/elasticsearch/action/admin/cluster/configuration/TransportAddVotingTombstonesAction.java

```diff
@@ -0,0 +1,160 @@
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Setting.Property;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.threadpool.ThreadPool.Names;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+public class TransportAddVotingTombstonesAction extends TransportMasterNodeAction<AddVotingTombstonesRequest, AddVotingTombstonesResponse> {
+
+    public static final Setting<Integer> MAXIMUM_VOTING_TOMBSTONES_SETTING
+        = Setting.intSetting("cluster.max_voting_tombstones", 10, 1, Property.Dynamic, Property.NodeScope);
+
+    @Inject
+    public TransportAddVotingTombstonesAction(TransportService transportService, ClusterService clusterService, ThreadPool threadPool,
+                                              ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver) {
+        super(AddVotingTombstonesAction.NAME, transportService, clusterService, threadPool, actionFilters, AddVotingTombstonesRequest::new,
+            indexNameExpressionResolver);
+    }
+
+    @Override
+    protected String executor() {
+        return Names.SAME;
+    }
+
+    @Override
+    protected AddVotingTombstonesResponse newResponse() {
+        throw new UnsupportedOperationException("usage of Streamable is to be replaced by Writeable");
+    }
+
+    @Override
+    protected AddVotingTombstonesResponse read(StreamInput in) throws IOException {
+        return new AddVotingTombstonesResponse(in);
+    }
+
+    @Override
+    protected void masterOperation(AddVotingTombstonesRequest request, ClusterState state,
+                                   ActionListener<AddVotingTombstonesResponse> listener) throws Exception {
+
+        request.resolveNodes(state); // throws IllegalArgumentException if no nodes matched
+
+        clusterService.submitStateUpdateTask("add-voting-tombstones", new ClusterStateUpdateTask() {
+
+            private Set<DiscoveryNode> resolvedNodes;
+
+            @Override
+            public ClusterState execute(ClusterState currentState) {
+                assert resolvedNodes == null : resolvedNodes;
+                resolvedNodes = request.resolveNodes(currentState);
+
+                final int oldTombstoneCount = currentState.getVotingTombstones().size();
+                final int newTombstoneCount = resolvedNodes.size();
+                final int maxTombstoneCount = MAXIMUM_VOTING_TOMBSTONES_SETTING.get(currentState.metaData().settings());
+                if (oldTombstoneCount + newTombstoneCount > maxTombstoneCount) {
```

Comment:
should we make this part of the pre-flight check as well?

Reply:
Ok, ab72d36170bc746b94cd560f6a701130749beaa1.